apps/ios/SimpleX SE/SEChatState.swift

```diff
@@ -31,34 +31,52 @@ class SEChatState {
 final class EventLoop {
     static let shared = EventLoop()
 
+    enum EventHandling {
+        case directMessage
+        case group
+    }
+
+    private var eventHandling: EventHandling?
     private var itemId: ChatItem.ID?
     private weak var model: ShareModel?
 
-    func set(itemId: ChatItem.ID, model: ShareModel) {
+    func set(handling: EventHandling, itemId: ChatItem.ID, model: ShareModel) {
+        self.eventHandling = handling
         self.itemId = itemId
         self.model = model
     }
 
     init() {
         Task {
             while true {
-                switch recvSimpleXMsg() {
+                let message = recvSimpleXMsg()
+                switch message {
+                // Drive the progress bar for direct chats
                 case let .sndFileProgressXFTP(_, aChatItem, _, sentSize, totalSize):
-                    if let id = aChatItem?.chatItem.id, id == itemId {
-                        await MainActor.run {
-                            withAnimation {
-                                model?.bottomBar = .loadingBar(progress: Double(sentSize) / Double(totalSize))
-                            }
+                    logger.debug("Received sndFileCompleteXFTP message, sentSize: \(sentSize), totalSize: \(totalSize)")
+                    guard eventHandling == .directMessage, let id = aChatItem?.chatItem.id, id == itemId else { continue }
+                    await MainActor.run {
+                        withAnimation {
+                            model?.bottomBar = .loadingBar(progress: Double(sentSize) / Double(totalSize))
                         }
                     }
                 case let .sndFileCompleteXFTP(_, aChatItem, _):
-                    if aChatItem.chatItem.id == itemId {
-                        await MainActor.run {
-                            model?.bottomBar = .loadingBar(progress: 1)
-                            model?.completion!(nil)
+                    logger.debug("Received sndFileCompleteXFTP message")
+                    guard let eventHandling, aChatItem.chatItem.id == itemId else { continue }
+                    switch eventHandling {
+                    case .directMessage:
+                        await MainActor.run { model?.bottomBar = .loadingBar(progress: 1) }
+                        // Dismiss the sheet only after completion has been received
+                        if aChatItem.chatItem.meta.itemStatus == .sndSent(sndProgress: .complete) {
+                            model?.completion?(nil)
                         }
+                    case .group:
+                        // For groups item status is not item status is likely to never be complete
+                        // Simply spin for 5 seconds
+                        try? await Task.sleep(for: .seconds(5))
```

Comment:
For groups, you need to wait for 3 things:
- sendMessage returns (it may happen before or after message sent or file complete event)
- message sent event arrives (with partial status)
- file complete event

You need to wait for at least 5 seconds after message sent event, provided file complete happened and sendMessage returned. But message sent event can arrive before, and this is currently not tracked yet.